File: 卡码/代码随想录/1/06-二叉树/Python/01-二叉树遍历方式/LeetCode_0515.py
#515. 在每个树行中找最大值
from typing import *
from ..TreeNode import *

class Solution:
  def largestValues(self, root: Optional[TreeNode]) -> List[int]:
    res:List[int]=[]
    q:List[TreeNode]=[]
    if root!=None:
      q.append(root)
    while len(q)>0:
      size:int=len(q)
      maxVal:int=q[0].val
      for i in range(size):
        node:TreeNode=q[0]
        q=q[1:]
        maxVal=node.val if node.val>maxVal else maxVal
        if node.left!=None:
          q.append(node.left)
        if node.right!=None:
          q.append(node.right)
      res.append(maxVal)
    return res

File: 卡码/代码随想录/Java/codeThoughts/src/main/java/com/kurumi/leetCode/codeThoughts/二叉树/_01_二叉树遍历方式/LeetCode_0515.java
package com.kurumi.leetCode.codeThoughts.二叉树._01_二叉树遍历方式;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

import com.kurumi.leetCode.codeThoughts.entity.TreeNode;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 515. 在每个树行中找最大值
 */
public class LeetCode_0515 {
    static class Solution {
        public List<Integer> largestValues(TreeNode root) {
            List<Integer> res = new ArrayList<>();
            Queue<TreeNode> q = new LinkedList<>();
            if (root != null)
                q.offer(root);
            while (!q.isEmpty()) {
                int size = q.size();
                int max = q.peek().val;
                while (size-- > 0) {
                    TreeNode node = q.poll();
                    max = max > node.val ? max : node.val;
                    if (node.left != null)
                        q.offer(node.left);
                    if (node.right != null)
                        q.offer(node.right);
                }
                res.add(max);
            }
            return res;
        }
    }
}
